### PR TITLE
Implemented BlockDataMock::getAsString(boolean)

### PR DESCRIPTION
**Pull request by svaningelgem**
# Description
In my code I was using the getAsString(true) method of BlockData. So, implementing it here.

# Checklist
The following items should be checked before the pull request can be merged.
- [x] Code follows existing style.
- [x] Unit tests added (if applicable).

### CODE REVIEW

**Comment by svaningelgem** [comment]
Hi,

Because of the modifications I made in here, the following tests start to fail:
```
    AmethystClusterMockTest. getAsString()
    BedMockTest. getAsString()
    CampfireMockTest. getAsString()
    SlabMockTest. getAsString()
    SwitchMockTest. getAsString()
    WallSignMockTest. getAsString()
```

But as far as I can see, my implementation is more correct than before.
Could you please double check?

If you agree with me, I'll adjust the tests mentioned above to call `getAsString(true)` instead.

**Comment by Thorinwasher** [comment]
> Could you please double check?

Best way for these issues (whenever the implementation seems too complex to just imitate) is to test how this works on a real instance. So simply just make a test plugin for a real paper instance (latest version) and then compare the real and the mock behavior. Alternatively you can of course just look at the paper source code, but I have a feeling that it's going to be complicated, as we're talking about serialization.

**Comment by svaningelgem** [comment]
Ok, I wrote a small plugin to dump this information.

This is the result:
```
[14:36:33 INFO]: [MaterialLogger] false|minecraft:yellow_wall_banner[facing=north]
[14:36:33 INFO]: [MaterialLogger] true|minecraft:yellow_wall_banner
[14:36:33 INFO]: [MaterialLogger] false|minecraft:yellow_wool
[14:36:33 INFO]: [MaterialLogger] true|minecraft:yellow_wool
[14:36:33 INFO]: [MaterialLogger] false|minecraft:zombie_head[powered=false,rotation=0]
[14:36:33 INFO]: [MaterialLogger] true|minecraft:zombie_head
[14:36:33 INFO]: [MaterialLogger] false|minecraft:zombie_wall_head[facing=north,powered=false]
[14:36:33 INFO]: [MaterialLogger] true|minecraft:zombie_wall_head
```

**Comment by svaningelgem** [comment]
I fooled a bit around and got this:
https://github.com/AvarionMC/dump_materials/releases/latest/download/material_data.json

Fully automated material generator :-). I can integrate this into MockBukkit? Or you can copy it in your own repos (I can transfer ownership)?

**Comment by Thorinwasher** [comment]
> I fooled a bit around and got this:

Good that you got a test plugin up and running, I don't see any need to have this in the mockbukkit repo, but it is still good to have as comparison for this pull request.

**Comment by svaningelgem** [comment]
FYI:
```
java.lang.IllegalArgumentException: Could not parse data: minecraft:chest[waterlogged=true,bumba=ABC]
Caused by: com.mojang.brigadier.exceptions.CommandSyntaxException: Block minecraft:chest does not have property 'bumba' at position 33: ...gged=true,<--[HERE]
```
==> So this means that paper can't understand invalid properties. (Good that I coded them in already 😁)

**Comment by svaningelgem** [comment]
FYI:
```
[20:51:12 INFO]: [MaterialLogger] Material: minecraft:diamond
[20:51:12 ERROR]: [MaterialLogger]  --> cannot create block data!
[20:51:12 INFO]: [MaterialLogger] Material: minecraft:diamond_axe
[20:51:12 ERROR]: [MaterialLogger]  --> cannot create block data!
[20:51:12 INFO]: [MaterialLogger] Material: minecraft:diamond_block
[20:51:12 INFO]: [MaterialLogger] false|minecraft:diamond_block
[20:51:12 INFO]: [MaterialLogger] true|minecraft:diamond_block
[20:51:12 INFO]: [MaterialLogger] Material: minecraft:diamond_boots
[20:51:12 ERROR]: [MaterialLogger]  --> cannot create block data!
```
So, from some materials, you can't create blocks.

However, that's where I'm stuck at atm:
```java
	{
		MaterialData materialData = new MaterialData(Material.DIAMOND);
		endermanMock.setCarriedMaterial(materialData);

		assertEquals(materialData, endermanMock.getCarriedMaterial());
	}
```

Here a `MaterialData` is created. Which underneath creates: `carriedBlock = BlockDataMock.mock(material.getItemType());`

However, we can't create this... (see above: we can't create a diamond blockdata even in pure paper.)

But for example an enderman can drop enderpearls. But that's not a valid material to create a block from... 

I'd like some guidance for this.
(and I'll push my changes for now so you can already verify what I changed).

**Comment by Thorinwasher** [comment]
> So, from some materials, you can't create blocks.

That's true, there's a method `Material#isBlock()` which you can use to determine this in mockbukkit. I hope that helps

**Comment by svaningelgem** [comment]
@Thorinwasher : This PR is ready (enough) for review :)

File: src/main/resources/materials/material_data.json
Comment (Thorinwasher) [inline src/main/resources/materials/material_data.json:1]: If this is being used in production, could you make a pull request to to MetaMiner with the script you made to retrieve this. (It's in the same organisation as MockBukkit)

**Review by Thorinwasher — DISMISSED** [review]
~Looks good to me (Please see the commnet below)~

I found one possible improvement

File: src/main/resources/materials/material_data.json
Comment (Thorinwasher) [inline src/main/resources/materials/material_data.json:1]: Split this into smaller files; one for each material would be good I believe (it's a lot of data here, but I'm interested if that would be better)

Comment (svaningelgem) [reply src/main/resources/materials/material_data.json:1]: I would suggest not to do that. There is too much data in there. And loading a json file won't take that long on modern systems. So unless there is some kind of grouping I could do, I'd like to keep it together...

Comment (svaningelgem) [reply src/main/resources/materials/material_data.json:1]: Done in PR1 there.

**Review by Thorinwasher — APPROVED** [review]

**Pull request merged**